build/stylecop.json

```diff
@@ -1,9 +1,10 @@
-﻿{
+{
   "$schema": "https://raw.githubusercontent.com/DotNetAnalyzers/StyleCopAnalyzers/master/StyleCop.Analyzers/StyleCop.Analyzers/Settings/stylecop.schema.json",
   "settings": {
     "documentationRules": {
       "copyrightText": "Copyright The OpenTelemetry Authors\nSPDX-License-Identifier: Apache-2.0",
-      "xmlHeader": false
+      "xmlHeader": false,
+      "documentInternalElements": false
```

Comment:
Why would want it now (what's the benefit vs. downside) considering "disable SA1602" hasn't been used before except for https://github.com/open-telemetry/opentelemetry-dotnet/pull/5255.

Reply:
e.g. these are working well

* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MetricPointStatus.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/AggregationType.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Prometheus.HttpListener/Internal/PrometheusType.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.GrpcNetClient/StatusCanonicalCode.cs

Reply:
This is a good change!

Before this PR, this code...

```csharp
// Generates no warnings whatsoever
internal class SomeClass
{
    public void Method()
    {
    }
}

internal enum SomeEnum
{
    Item1, // Generates SA1602
    Item2, // Generates SA1602
    Item3, // Generates SA1602
}
```

The reason for that is: https://github.com/open-telemetry/opentelemetry-dotnet/blob/914ebf9c6d369803d1754c4c5bad67874ca326fb/build/OpenTelemetry.prod.ruleset#L12

If we remove that rule tweak for SA1600 the code ends up like this:

```csharp
internal class SomeClass
{
    public void Method() // Generates SA1600
    {
    }
}

internal enum SomeEnum
{
    Item1, // Generates SA1602
    Item2, // Generates SA1602
    Item3, // Generates SA1602
}
```

So we intentionally disabled the warning for some internal things, but not others. Really it was disabled in the incorrect way.

What this PR is doing is treating enum items the same way we treat other internal things.

The other ways to deal with the warning...

* Suppress it:

   ```csharp
   internal enum SomeEnum
   {
   #pragma warning disable SA1602 // Enumeration items should be documented
       Item1,
       Item2,
       Item3,
   #pragma warning restore SA1602 // Enumeration items should be documented
   }
   ```

   This is ugly. Generates cruft in the code.

* Add the docs:

   ```csharp
   internal enum SomeEnum
   {
    /// <summary>
    /// Item1.
    /// </summary>
    Item1,

    /// <summary>
    /// Item2.
    /// </summary>
    Item2,

    /// <summary>
    /// Item3.
    /// </summary>
    Item3,
   }
   ```

   Doesn't really add any value IMO. Just ceremony/bloat 🤷

PS: Having uncovered this, I'm going to push an update to clean up the ruleset as well.

Reply:
I could add docs to the PR in question, but it seemed overkill. This seems pretty self-explanatory to me:

```csharp
    internal enum LogRecordSource
    {
        CreatedManually,
        FromThreadStaticPool,
        FromSharedPool,
    }
```

vs

```csharp
    internal enum LogRecordSource
    {
        /// <summary>
        /// A LogRecord created manually.
        /// </summary>
        CreatedManually,

        /// <summary>
        /// A LogRecord rented from the static pool.
        /// </summary>
        FromThreadStaticPool,

        /// <summary>
        /// A LogRecord rented from the shared pool.
        /// </summary>
        FromSharedPool,
    }
```

Agree or no?

Reply:
> Agree or no?

I agree this specific example is quite self-explanatory. I disagree that this is a common case that should be applied everywhere.

Reply:
Can you expand on why you feel we should treat enum items differently than other public internal APIs?

Reply:
Here are some internal artifacts which do not have documentation (I just selected a few more or less at random, there are many):

* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Internal/SelfDiagnosticsConfigParser.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Internal/WildcardHelper.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/AggregatorStore.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/LookupData.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MetricPointValueStorage.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MetricStreamIdentity.cs

Lots of things with mixed/partial docs:

* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/Base2ExponentialBucketHistogram.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/HistogramBuckets.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MeterProviderSdk.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MetricPointOptionalComponents.cs
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MetricReaderExt.cs

My question is what is so special about internal enum items that we enforce those to be documented but not all internal code? I feel the reverse would be a more sensible stance, those enums are a lot more self-explanatory than this code 🤣

Reply:
> My question is what is so special about internal enum items that we enforce those to be documented but not all internal code? I feel the reverse would be a more sensible stance, those enums are a lot more self-explanatory than this code 🤣

I don't think enums are special. For internal code, I think they should be documented. I don't see why we would want to keep lowing the bar without a good reason.

Reply:
I don't see it as lowering. I see it as correcting an oversight, correcting an inconsistency. The intention was to NOT require docs for internal members, because they are burdensome. Code should be well-written, not reliant on docs for clarity. It is a code review concern 😄 If we see an enum which needs docs to be clear, that is a code smell.

https://github.com/search?q=repo%3Adotnet%2Fruntime+%22internal+enum%22&type=code

Reply:
> Code should be well-written, not reliant on docs for clarity.

I agree. Would you clean up the docs for internal members (e.g. https://github.com/open-telemetry/opentelemetry-dotnet/pull/5256#discussion_r1466905651) and see if that makes things better. If most of them are self-explanatory, that's a good reason to flip the default behavior.